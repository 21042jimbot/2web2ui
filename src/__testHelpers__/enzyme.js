export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
