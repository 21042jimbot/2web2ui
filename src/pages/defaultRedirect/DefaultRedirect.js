import React, { Component } from 'react';
import _ from 'lodash';
import { Loading } from 'src/components/loading/Loading';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import selectAccessConditionState from 'src/selectors/accessConditionState';
import config from 'src/config';

export class DefaultRedirect extends Component {

  componentDidMount() {
    this.handleRedirect();
  }

  componentDidUpdate() {
    this.handleRedirect();
  }

  handleRedirect() {
    const { location, history, currentUser, ready } = this.props;
    const { state: routerState = {}} = location;
    const allowedAccessLevels = ['reporting', 'heroku', 'azure'];

    // if there is a redirect route set on state, we can
    // redirect there before access condition state is ready
    if (routerState.redirectAfterLogin) {
      history.push({ ...location, pathname: routerState.redirectAfterLogin }, { replace: true });
      return;
    }

    // if access condition state hasn't loaded, we can't
    // make a redirect decision yet
    if (!ready) {
      return;
    }

    // reporting users are all sent to the summary report
    if (_.includes(allowedAccessLevels, currentUser.access_level)) {
      history.push({ ...location, pathname: '/reports/summary' }, { replace: true });
      return;
    }

    // everyone else is sent to the config.splashPage route
    history.push({ ...location, pathname: config.splashPage }, { replace: true });
  }

  render() {
    return <Loading />;
  }
}

const mapStateToProps = (state) => selectAccessConditionState(state);
export default withRouter(connect(mapStateToProps)(DefaultRedirect));
