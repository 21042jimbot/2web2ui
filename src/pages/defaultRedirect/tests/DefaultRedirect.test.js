import React from 'react';
import { shallow } from 'enzyme';
import { DefaultRedirect } from '../DefaultRedirect';

jest.mock('src/config', () => ({
  splashPage: '/config/splash'
}));

describe('Component: DefaultRedirect', () => {

  let props;
  let wrapper;
  let instance;

  beforeEach(() => {
    props = {
      location: {
        state: {},
        search: '?test=one'
      },
      history: {
        push: jest.fn(),
        replace: jest.fn()
      },
      currentUser: {
        access_level: 'admin'
      },
      ready: false
    };
    props.history.push.mockName('historyPushMock');
    props.history.replace.mockName('historyReplaceMock');
    wrapper = shallow(<DefaultRedirect {...props} />);
    instance = wrapper.instance();
  });

  it('should render correctly by default', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should redirect after login', () => {
    wrapper.setProps({ location: { state: { redirectAfterLogin: '/test/redirect/after/login' }}});
    props.history.replace.mockClear();
    instance.handleRedirect();
    expect(props.history.replace).toHaveBeenCalledTimes(1);
    expect(props.history.replace).toHaveBeenCalledWith({
      pathname: '/test/redirect/after/login',
      state: { redirectAfterLogin: '/test/redirect/after/login' }
    });
  });

  it('should do nothing if no redirect after login and not ready', () => {
    props.history.replace.mockClear();
    instance.handleRedirect();
    expect(props.history.replace).not.toHaveBeenCalled();
  });

  it('should redirect to summary report for reporting users', () => {
    wrapper.setProps({
      currentUser: {
        access_level: 'reporting'
      },
      ready: true
    });
    props.history.replace.mockClear();
    instance.handleRedirect();
    expect(props.history.replace).toHaveBeenCalledTimes(1);
    expect(props.history.replace).toHaveBeenCalledWith({
      pathname: '/reports/summary',
      search: '?test=one',
      state: {}
    });
  });

  it('should redirect based on config', () => {
    wrapper.setProps({ ready: true });
    props.history.replace.mockClear();
    instance.handleRedirect();
    expect(props.history.replace).toHaveBeenCalledTimes(1);
    expect(props.history.replace).toHaveBeenCalledWith({
      pathname: '/config/splash',
      search: '?test=one',
      state: {}
    });
  });

  it('should handle a redirect on mount', () => {
    jest.spyOn(instance, 'handleRedirect');
    instance.componentDidMount();
    expect(instance.handleRedirect).toHaveBeenCalledTimes(1);
  });

  it('should handle a redirect on update', () => {
    jest.spyOn(instance, 'handleRedirect');
    instance.componentDidUpdate();
    expect(instance.handleRedirect).toHaveBeenCalledTimes(1);
  });

});
