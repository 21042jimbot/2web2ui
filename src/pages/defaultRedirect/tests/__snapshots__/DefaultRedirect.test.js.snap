// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultRedirect should render correctly by default 1`] = `<Loading />`;
